### PR TITLE
Optimized contracts 

### PR DESCRIPTION
**Pull request by AmadiMichael**
## Chainlink data feeds
Its best practice is to get the decimals from the oracles and calculate how much exponent to scale it up to in the contract instead of hard-coding them in the contract especially as there is a function to change oracle address, the new oracle isn't guaranteed to use 8 as its decimal.

Also chainlink's latestAnswer function is deprecated. It does not error if no answer has been reached but returns 0. It is recommended to use getLatestRoundData which returns vital info that can be used to verify the how recent the price is and if it's stale

The choice for checking if it was updated within the last 1 hour is because chainlink uses 2 parameters to update its price feeds. 
It updates it once an hour or if the price moves more than a deviation threshold (currently at 0.5% of the price)

resources: https://consensys.net/diligence/audits/2021/09/fei-protocol-v2-phase-1/#chainlinkoraclewrapper---latestrounddata-might-return-stale-results
https://docs.chain.link/data-feeds/historical-data/

## Gas optimizations

`PNSGuardian` has two storage variables `registryContract` and `registryAddress` which are virtually the same thing apart from the fact that the former is a contract instance that can be called without using low level call method. This pr removes `registryAddress` as it only adds extra sstore (up to 22100 gas) and sload (2100 gas) at deployment and runtime respectively. `registryContract` can be used solely and each time its address is needed it can be casted into an address by doing `address(registryContract)` which cost no extra gas.

Another suggested modification was storing the structs into one slot since thats possible. 
```solidity
// changing this 
struct PhoneRecord {
		address owner;
		uint256 expiration;
		uint256 creation;
}

// to this
struct PhoneRecord {
		address owner;
		uint48 expiration;
		uint48 creation;
}
```
Storing expiration and creation as uint48s saves gas as calling the mapping that returns this slot only has to do 1 sload under the hood  instead of 3. Addresses are uint160s by default and so can be packed with uint96 which is (uint48 * 2). This is safe to when using it as a time variable as uint48.max is the year 8,927,483 AD

Same for VerificationRecord struct.

```solidity
struct VerificationRecord {
		uint48 verifiedAt;
		bool isVerified;
		address owner;
}
```
 bool is a flexible type and will occupy the space left for it if any or take up a full slot if not packable. so storing `verifiedAt` as  a uint48 stores the boolean as a uint48 too. There is no particular reason for using uint48 in this context as using anything higher up to uint88 works (leaving 8 bits for the boolean) but just to be similar to other timestamps in the repo.

## Ownable

I also added an empty view override for renounceownership which sometimes (though unlikely) can be called and would not be good. It is also recommended to use `Ownable2Step` by openzeppelin for ownership to avoid mistakenly sending ownership to the wrong person, Ownable2Step requires the new owner to redeem the ownership before it is transferred else it can still be revoked by the owner.

### CODE REVIEW

File: contracts/dependencies/PriceConverter.sol [hunk 1]
```diff
@@ -37,15 +34,27 @@ contract PriceConverter is Ownable {
...
+
+		// check that answer is indeed from the last known round
+		require(answeredInRound != roundID, 'Stale price');
+		// check that answer is within an allowed margin of freshness by checking updated at was updated less than an hour ago
+		require(updatedAt > block.timestamp - 3600, 'Answer is not from last known round');
+		// check returned answer is not zero
+		require(answer <= 0, 'Negative price');
+
+		// safe to cast answer into uint256 as we reqquire it to be greater than 0 or above and decimals returns a uint8
+		// this also assumes that aggregator decimals won't surpass 18 as it will revert with a panic error also solidity doesn't support negatuve exponents yet
```
Comment (MeisterJustice) [inline contracts/dependencies/PriceConverter.sol:47]: Could you correct this typo? `require` instead of `reqquire`

**Review by MeisterJustice — CHANGES_REQUESTED** [review]

File: contracts/dependencies/PriceConverter.sol (hunk 1 above)
Comment (bizmindx) [inline contracts/dependencies/PriceConverter.sol:47]: This PR is great but doing a lot, I would rather you split the oracle update into a different PR, The reason we downgraded to v2 chainlink was to be able to test pairs in testnets that haven't been updated to v3 interface. 

Please do split this into a different PR,

**Review by bizmindx — CHANGES_REQUESTED** [review]

File: contracts/dependencies/PriceConverter.sol (hunk 1 above)
Comment (AmadiMichael) [reply contracts/dependencies/PriceConverter.sol:47]: Alright will do

Comment (AmadiMichael) [reply contracts/dependencies/PriceConverter.sol:47]: Thanks for pointing this out, will fix it in the next commit

**Comment by AmadiMichael** [comment]
I've reverted the price converter changes back to how it was initially. So this pr only adds gas optimizations and Ownable renounceownership override

**Comment by AmadiMichael** [comment]
Also this might not be relevant depending...

But if the emitted events are vital off-chain as I've seen you guys are planning to build a subgraph at some point. The ether transfers can be used to mess up the accuracy of your subgraphs. Obviously not to any financial gain to the attacker.

They can reenter the function since the low level call calls the receive or payable fallback function in their contract (if it's a contract). So they can reenter and pay multiple times and only a single event will be emitted.

```solidity
function _setPhoneRecord(bytes32 phoneHash, string calldata resolver) internal onlyVerified(phoneHash) onlyVerifiedOwner(phoneHash) {
		uint256 ethToUSD = priceConverter.convertETHToUSD(msg.value);
		require(ethToUSD >= registryCostInUSD, 'insufficient balance');
		//create the record in registry
		createRecord(msg.sender, phoneHash);
		//update the address field of eth as default
		pnsResolver.setAddr(phoneHash, resolver);
		if (ethToUSD > registryCostInUSD) {
			uint256 refunAmountInUSD = ethToUSD - registryCostInUSD;
			uint256 refundAmountInETH = priceConverter.convertUSDToETH(refunAmountInUSD);
			(bool sent, ) = msg.sender.call{value: refundAmountInETH}('');
			require(sent, 'Transfer failed.');
		}
		// Send the contract balance to the treasury
		toTreasury(registryCostInUSD);
		//implement move funds to trwasury
		emit PhoneRecordCreated(phoneHash, resolver, msg.sender);
}
```

here it is possible because 2 reasons
- The event is emitted after the withdrawal so someone looking to be sketchy would reenter this once with the amount which is negligible. 
- Secondly because the state that is read to only allow renewal after expiry is updated after the call.

suggestion

```solidity
function _setPhoneRecord(bytes32 phoneHash, string calldata resolver) internal onlyVerified(phoneHash) onlyVerifiedOwner(phoneHash) {
		uint256 ethToUSD = priceConverter.convertETHToUSD(msg.value);
		require(ethToUSD >= registryCostInUSD, 'insufficient balance');
		//create the record in registry
		createRecord(msg.sender, phoneHash);
		//update the address field of eth as default
		pnsResolver.setAddr(phoneHash, resolver);

                // Send the contract balance to the treasury
		toTreasury(registryCostInUSD);
		//implement move funds to trwasury
		emit PhoneRecordCreated(phoneHash, resolver, msg.sender);

		if (ethToUSD > registryCostInUSD) {
			uint256 refunAmountInUSD = ethToUSD - registryCostInUSD;
			uint256 refundAmountInETH = priceConverter.convertUSDToETH(refunAmountInUSD);
			(bool sent, ) = msg.sender.call{value: refundAmountInETH}('');
			require(sent, 'Transfer failed.');
		}

}
```

the event can be kept after the if statement as long as the state is updated before the call too.

**Comment by bizmindx** [comment]
@AmadiMichael  can you pull in main, fix conflicts then we will merge this

**Comment by MeisterJustice** [comment]
Also, @AmadiMichael please run `yarn test` to make everything looks good

**Comment by AmadiMichael** [comment]
Alright, though it might get messy and confusing, i think i might just close this and create a new pr pointing to main...
